README.md

Requirements:
 - MQTT v3 (or higher) broker with support for retained messages
 - It is preferred to send messages to yeelight2mqtt with QoS 2, to avoid Yeelight's rate limiting. 

CHANGELOG.md
# Changelog

## [0.2.0](https://github.com/dsorm/yeelight2mqtt/compare/v0.1.0...v0.2.0) (2023-03-16)


### Features

* added build variables ([05c5706](https://github.com/dsorm/yeelight2mqtt/commit/05c5706aa8af5c6b15bb38017f553e4a3f994453))
* improve go workflow ([ce3ede3](https://github.com/dsorm/yeelight2mqtt/commit/ce3ede3d5276a4d462655e23db35ff643d615fc2))
* improved and simplified Dockerfile ([65d06a8](https://github.com/dsorm/yeelight2mqtt/commit/65d06a8884f07742ef769d173fb3a5fea96fbac7))
* try improving the go action ([2c1129e](https://github.com/dsorm/yeelight2mqtt/commit/2c1129e126ca6fb7f960714dc8f02dc977276148))
* try using matrixes in go action ([6769ada](https://github.com/dsorm/yeelight2mqtt/commit/6769ada769687a1c601461fdba96878612359c6e))
* update docker.yml ([e19eea8](https://github.com/dsorm/yeelight2mqtt/commit/e19eea8e5f4703bca3c85b940aa83f7abccfd680))
* update docker.yml ([1075df0](https://github.com/dsorm/yeelight2mqtt/commit/1075df008eb94aa52db9f35351fda20d9829494d))
* Update docker.yml ([ac6d01e](https://github.com/dsorm/yeelight2mqtt/commit/ac6d01ee11592e004af4a0bce793dc9b82dd0e02))


### Bug Fixes

* add download dependencies step ([c16b973](https://github.com/dsorm/yeelight2mqtt/commit/c16b973beb06f57ecf0ebef7c941d7ef55a437f8))
* change go ver ([6cc9abf](https://github.com/dsorm/yeelight2mqtt/commit/6cc9abf771d20322728d34b05d7952bcce88f7e9))
* error ([37bb5a1](https://github.com/dsorm/yeelight2mqtt/commit/37bb5a122a77725ed5746e4ce36397abe4e91963))
* remove go get . ([64b9080](https://github.com/dsorm/yeelight2mqtt/commit/64b90800bef8ec1492bb78b6eb0129e82cf69c35))
* remove String() funcs so the tests pass ([9324c59](https://github.com/dsorm/yeelight2mqtt/commit/9324c596847a1a0bafc490cfdb37f0595cba66b1))
* tag in actions workflow ([054a754](https://github.com/dsorm/yeelight2mqtt/commit/054a75422d55a4c8c4aaf7ec0ebc228170e9bf34))
* update ([bd73e84](https://github.com/dsorm/yeelight2mqtt/commit/bd73e84fba7a1fd1d5dabb00ae2e79c03c252692))

## [0.1.0](https://github.com/dsorm/yeelight2mqtt/compare/v0.0.1...v0.1.0) (2023-03-16)


### Features

* add release please ([91a7d72](https://github.com/dsorm/yeelight2mqtt/commit/91a7d721894cc3f03299727a08253f229059d175))


### Bug Fixes

* get dependencies in action ([1c8282b](https://github.com/dsorm/yeelight2mqtt/commit/1c8282bbd2ddaf2cc8aac6912764b722337ad68d))
* rename file and bump go version ([e820004](https://github.com/dsorm/yeelight2mqtt/commit/e8200043989d2a898564c06fe0a63f667d17803d))


api/light.go
package api

import (
	"net"
	"sync"
)

type Light struct {
	Host string
	Name string

	stateMutex  sync.Mutex
	latestState LightProperties
	conn        net.Conn
	connMutex   sync.Mutex

	// When there is a change in Yeelight props, it is automatically sent by the light back to yeelight2mqtt
	// yeelight2mqtt only reads from the connection when a request is made or because of automatic polling,
	// so this channel is used to get rid of the data (when handling another request), so a dedicated goroutine may handle it
	refreshMsg      chan string
	refreshCallback func(message string)
}

type LightProperties struct {
	On             bool      // "on" or "off"
	Bright         uint8     // (range 1 - 100)
	Ct             uint16    // (range 1700 - 6500) (unit: Kelvin)
	RGB            uint32    // (range 0 - 16777215)
	Hue            uint16    // (range 0 - 359)
	Sat            uint8     // (range 0 - 100)
	Color_Mode     ColorMode // 0: RGB mode, 1: Color temperature mode, 2: HSV mode, 3: Flow mode
	Flowing        bool
	Delayoff       uint8 // (range 1 - 60 minutes)
	Flow_Params    string
	Music_On       bool
	Name           string
	Bg_On          bool
	Bg_Flowing     bool
	Bg_Flow_Params string
	Bg_Ct          uint16
	Bg_Color_Mode  ColorMode
	Bg_Bright      uint8
	Bg_RGB         uint32
	Bg_Hue         uint16
	Bg_Sat         uint8
	Nl_Br          uint8 // (range 1 - 100)
	Moonlight_On   bool
}

type ColorMode uint8

const (
	ColorModeRGB ColorMode = iota
	ColorModeCT
	ColorModeHSV
	ColorModeFlow
)

func (cm ColorMode) String() string {
	switch cm {
	case ColorModeRGB:
		return "RGB"
	case ColorModeCT:
		return "CT"
	case ColorModeHSV:
		return "HSV"
	case ColorModeFlow:
		return "Flow"
	}
	return ""
}

func ColorModeFromString(str string) (ColorMode, error) {
	switch str {
	case "RGB":
		return ColorModeRGB, nil
	case "CT":
		return ColorModeCT, nil
	case "HSV":
		return ColorModeHSV, nil
	case "Flow":
		return ColorModeFlow, nil
	}
	return 0, nil
}


api/sendCommand.go
package api

import (
	"errors"
	"fmt"
	"github.com/dsorm/yeelight2mqtt/console"
	"net"
	"strings"
	"time"
)

// for debug only, temporary
var successfulSends uint64
var unsuccessfulSends uint64

type sendCommandCtx struct {
	tries    uint64
	maxTries uint64
	command  string
	l        *Light
}

func (l *Light) SendCommand(command string, maxTries int) (response []byte, err error) {
	ctx := &sendCommandCtx{
		maxTries: uint64(maxTries),
		command:  command,
		l:        l,
	}
	return ctx.sendCommandWithCtx()
}

// this code is trash, and it probably doesn't even work properly
func (ctx *sendCommandCtx) sendCommandWithCtx() (response []byte, err error) {
	ctx.tries++
	if ctx.tries >= ctx.maxTries {
		return nil, errors.New("max tries exceeded")
	}

	ctx.l.connMutex.Lock()

	// a very dirty way to ratelimit the number of requests to the Yeelight, but it works :)
	done := make(chan bool)
	unlockMutex := func() {
		done <- true
		time.Sleep(time.Second / 2)
		ctx.l.connMutex.Unlock()
	}

	// this function serves as a force-unlock in case everything goes wrong (it will, trust me)
	go func() {
		timeoutSeconds := 40
		ticker := time.NewTicker(time.Second * time.Duration(timeoutSeconds))

		select {
		case <-ticker.C:
			console.Logf("sendCommand was stuck for %v seconds, force-unlocking mutex and closing connection\n", timeoutSeconds)
			err = ctx.l.conn.Close()
			if err != nil {
				console.Logf("error closing net connection: %v\n", err)
			}

			unlockMutex()
			return
		case <-done:
			ticker.Stop()
			return
		}

	}()

	if ctx.l.conn == nil {
		ctx.l.conn, err = net.DialTimeout("tcp", fmt.Sprintf("%s:55443", ctx.l.Host), 5*time.Second)
		if err != nil {
			unlockMutex()
			return ctx.sendCommandWithCtx()
		}
	}

	// wait a while before writing command, since yeelights are a bit slow,
	// they might write the output of last command to this command or just crap out
	time.Sleep(time.Second / 4)

	// if ctx.command is empty, just read from the connection
	if ctx.command != "" {
		_, err = ctx.l.conn.Write([]byte(ctx.command + "\r\n"))
		if err != nil {
			err = ctx.l.conn.Close()
			if err != nil {
				console.Logf("error closing net connection: %v\n", err)
			}
			ctx.l.conn = nil
			unlockMutex()
			ctx.tries++
			return ctx.sendCommandWithCtx()
		}

	}

	err = ctx.l.conn.SetDeadline(time.Now().Add(time.Second * 5))
	if err != nil {
		err = ctx.l.conn.Close()
		if err != nil {
			console.Logf("error closing net connection: %v\n", err)
		}
		ctx.l.conn = nil
		unlockMutex()
		return ctx.sendCommandWithCtx()
	}

	resp := []byte{}
	// read a line from response
	for {
		buf := make([]byte, 1024)
		n, err := ctx.l.conn.Read(buf)
		if err != nil {
			reason := "unknown reason"
			if strings.Contains(err.Error(), "timeout") {
				reason = "timeout"
			}
			console.Logf("sendCommand() to %v failed due to %v, retrying... (%v/%v)\n", ctx.l.Host, reason, ctx.tries, ctx.maxTries)

			unsuccessfulSends++
			unlockMutex()
			return ctx.sendCommandWithCtx()
		}
		resp = append(resp, buf[:n]...)
		if string(resp[len(resp)-2:]) == "\r\n" {
			successfulSends++
			break
		}
	}

	// detect if the yeelight is vomiting the (currently) unnecessary output about prop change
	if strings.HasPrefix(string(resp), "{\"method\":\"props\",\"params\":{") {
		// TODO make it work
		// ctx.l.refreshMsg <- string(resp)

		// read again to get the requested response
		ctx.tries -= 2
		ctx.command = ""
		unlockMutex()
		return ctx.sendCommandWithCtx()
	}

	go unlockMutex()
	return resp, err
}


api/commands.go
package api

import (
	"bytes"
	"encoding/json"
	"fmt"
	"github.com/dsorm/yeelight2mqtt/console"
	"log"
	"reflect"
	"strconv"
	"strings"
)

func (l *Light) sendVerify(funcName string, command string, params ...interface{}) error {
	formattedCommand := fmt.Sprintf(command, params...)
	response, err := l.SendCommand(formattedCommand, 10)

	if err != nil {
		return err
	}

	if !strings.Contains(string(response), "ok") {
		return fmt.Errorf("%v() failed:\n\tresponse from light: %v", funcName, string(response))
	}

	return nil
}

func (l *Light) GetProp() error {
	command := "{\"id\":0,\"method\":\"get_prop\",\"params\":[\"power\", \"bright\", \"ct\", \"rgb\", \"hue\", \"sat\", \"color_mode\", \"flowing\", \"delayoff\", \"flow_params\", \"music_on\", \"name\", \"bg_power\", \"bg_flowing\", \"bg_flow_params\", \"bg_ct\", \"bg_lmode\", \"bg_bright\", \"bg_rgb\", \"bg_hue\", \"bg_sat\", \"nl_br\", \"active_mode\"]}"
	resp, err := l.SendCommand(command, 3)

	if err != nil {
		return err
	}

	// var props lightProperties
	d := json.NewDecoder(bytes.NewReader(resp))

	var m map[string]interface{}
	err = d.Decode(&m)
	if err != nil {
		return err
	}

	// make sure m["result"] is not nill
	_, keyExists := m["result"]
	if !keyExists {
		return fmt.Errorf("GetProp() failed: output from light doesn't contain result")
	}

	// make sure m["result"] is the right type
	if reflect.TypeOf(m["result"]).Kind() != reflect.Slice {
		return fmt.Errorf("GetProp() failed: %v", m["result"])
	}
	result := m["result"].([]interface{})
	if len(result) < 21 {
		return fmt.Errorf("GetProp() failed: result too small (expected at least 21, got %v), result is %v\n", len(result), result)
	}

	atouint8 := func(i interface{}) uint8 {
		s := i.(string)
		if i.(string) == "" {
			return 0
		}

		num, err := strconv.Atoi(s)
		if err != nil {
			log.Fatalf("atoi: %v", err)
		}
		return uint8(num)
	}

	atouint16 := func(i interface{}) uint16 {
		s := i.(string)
		if i.(string) == "" {
			return 0
		}
		num, err := strconv.Atoi(s)
		if err != nil {
			console.Logf("atoi: %v\n", err)
		}
		return uint16(num)
	}

	atouint32 := func(i interface{}) uint32 {
		s := i.(string)
		if i.(string) == "" {
			return 0
		}

		num, err := strconv.Atoi(s)
		if err != nil {
			console.Logf("atoi: %v\n", err)
		}
		return uint32(num)
	}
	lp := LightProperties{
		On:             result[0].(string) == "on",
		Bright:         atouint8(result[1]),
		Ct:             atouint16(result[2]),
		RGB:            atouint32(result[3]),
		Hue:            atouint16(result[4]),
		Sat:            atouint8(result[5]),
		Color_Mode:     ColorMode(atouint8(result[6])),
		Flowing:        result[7].(string) == "1",
		Delayoff:       atouint8(result[8]),
		Flow_Params:    result[9].(string),
		Music_On:       result[10].(string) == "1",
		Name:           result[11].(string),
		Bg_On:          result[12].(string) == "on",
		Bg_Flowing:     result[13].(string) == "1",
		Bg_Flow_Params: result[14].(string),
		Bg_Ct:          atouint16(result[15]),
		Bg_Color_Mode:  ColorMode(atouint8(result[16])),
		Bg_Bright:      atouint8(result[17]),
		Bg_RGB:         atouint32(result[18]),
		Bg_Hue:         atouint16(result[19]),
		Bg_Sat:         atouint8(result[20]),
		Nl_Br:          atouint8(result[21]),
		Moonlight_On:   result[22].(string) == "1",
	}

	l.latestState = lp
	return nil
}

/*
"ct_value" is the target color temperature. The type is integer and
range is 1700 ~ 6500 (k).

"effect" support two values: "sudden" and "smooth".

	If effect is "sudden",  then the color temperature will be changed directly to target value,
	under this case, the third parameter "duration" is ignored.
	If effect is "smooth", then the color temperature will be changed to target value in a
	gradual fashion, under this case, the total time of gradual change is specified in third
	parameter "duration".

"duration" specifies the total time of the gradual changing.

	The minimum support duration is 30 milliseconds.

From Yeelight's Inter-operation Specification
*/
func (l *Light) SetCtAbx(ct_value uint, effect string, duration string) error {
	if ct_value < 1700 || ct_value > 6500 {
		return fmt.Errorf("SetCtAbx() failed: ct_value out of range")
	}
	if effect != "sudden" && effect != "smooth" {
		return fmt.Errorf("SetCtAbx() failed: effect must be 'sudden' or 'smooth'")
	}

	durationConv, err := strconv.Atoi(duration)
	if err != nil {
		return fmt.Errorf("SetCtAbx() failed: duration must be an integer")
	}
	if durationConv < 30 {
		return fmt.Errorf("SetCtAbx() failed: duration must be at least 30 ms")
	}

	err = l.sendVerify("SetCtAbx", "{\"id\":0,\"method\":\"set_ct_abx\",\"params\":[%v, \"%v\", %v]}", ct_value, effect, duration)
	if err != nil {
		return err
	}

	l.stateMutex.Lock()
	l.latestState.Ct = uint16(ct_value)
	l.stateMutex.Unlock()
	return nil
}

/*
"rgb_value" is the target color, whose type is integer.

	It should be expressed in decimal integer ranges from 0 to 16777215 (hex: 0xFFFFFF).
	RED, GREEN, BLUE = 0 up to 255
	rgb_value = RED*65536 + GREEN*256 + BLUE

"effect" support two values: "sudden" and "smooth".

	If effect is "sudden",  then the color temperature will be changed directly to target value,
	under this case, the third parameter "duration" is ignored.
	If effect is "smooth", then the color temperature will be changed to target value in a
	gradual fashion, under this case, the total time of gradual change is specified in third
	parameter "duration".

"duration" specifies the total time of the gradual changing.

	The minimum support duration is 30 milliseconds.

From Yeelight's Inter-operation Specification
*/
func (l *Light) SetRGB(rgb_value uint32, effect string, duration string) error {
	if rgb_value > 16777215 {
		return fmt.Errorf("SetRGB() failed: rgb_value out of range")
	}

	err := l.sendVerify("SetRGB", "{\"id\":0,\"method\":\"set_rgb\",\"params\":[%v, \"%v\", %v]}", rgb_value, effect, duration)
	if err != nil {
		return err
	}

	l.stateMutex.Lock()
	l.latestState.RGB = rgb_value
	l.stateMutex.Unlock()
	return nil
}

/*
"hue" is the target hue value, whose type is integer.

	It should be expressed in decimal integer ranges from 0 to 359.

"sat" is the target saturation value whose type is integer.

	It's range is 0 to 100.

"effect" support two values: "sudden" and "smooth".

	If effect is "sudden",  then the color temperature will be changed directly to target value,
	under this case, the third parameter "duration" is ignored.
	If effect is "smooth", then the color temperature will be changed to target value in a
	gradual fashion, under this case, the total time of gradual change is specified in third
	parameter "duration".

"duration" specifies the total time of the gradual changing.

	The minimum support duration is 30 milliseconds.

From Yeelight's Inter-operation Specification
*/
func (l *Light) SetHSV(hue uint16, sat uint8, effect string, duration string) error {
	if hue > 359 {
		return fmt.Errorf("SetHSV() failed: hue out of range")
	}
	if sat > 100 {
		return fmt.Errorf("SetHSV() failed: sat out of range")
	}

	err := l.sendVerify("SetHSV", "{\"id\":0,\"method\":\"set_hsv\",\"params\":[%v, %v, \"%v\", %v]}", hue, sat, effect, duration)
	if err != nil {
		return err
	}

	l.stateMutex.Lock()
	l.latestState.Hue = hue
	l.stateMutex.Unlock()
	return nil
}

/*
"brightness" is the target brightness.

	The type is integer and ranges from 1 to 100. The brightness is a
	percentage instead of a absolute value. 100 means maximum brightness
	while 1 means the minimum brightness.

"effect" support two values: "sudden" and "smooth".

	If effect is "sudden",  then the color temperature will be changed directly to target value,
	under this case, the third parameter "duration" is ignored.
	If effect is "smooth", then the color temperature will be changed to target value in a
	gradual fashion, under this case, the total time of gradual change is specified in third
	parameter "duration".

"duration" specifies the total time of the gradual changing.

	The minimum support duration is 30 milliseconds.

From Yeelight's Inter-operation Specification
*/
func (l *Light) SetBright(brightness uint8, effect string, duration string) error {
	if brightness < 1 || brightness > 100 {
		return fmt.Errorf("SetBright() failed: brightness out of range")
	}

	err := l.sendVerify("SetBright", "{\"id\":0,\"method\":\"set_bright\",\"params\":[%v, \"%v\", %v]}", brightness, effect, duration)
	if err != nil {
		return err
	}

	l.stateMutex.Lock()
	l.latestState.Bright = brightness
	l.stateMutex.Unlock()
	return nil
}

/*
		"power" can only be "on" or "off".
			"on" means turn on the smart LED,
	        "off" means turn off the smart LED.

		"effect" support two values: "sudden" and "smooth".
			If effect is "sudden", then the color temperature will be changed directly to target value,
			under this case, the third parameter "duration" is ignored.
			If effect is "smooth", then the color temperature will be changed to target value in a
			gradual fashion, under this case, the total time of gradual change is specified in third
			parameter "duration".

		"duration" specifies the total time of the gradual changing. The minimum support duration is
		30 milliseconds.

		"mode" (optional):
			0: Normal turn on operation (default value)
			1: Turn on and switch to CT mode.
			2: Turn on and switch to RGB mode.
			3: Turn on and switch to HSV mode.
			4: Turn on and switch to color flow mode.
	        5: Turn on and switch to Night light mode. (Ceiling light only).

		From Yeelight's Inter-operation Specification
*/
func (l *Light) SetPower(power string, effect string, duration string, mode string) error {
	if len(mode) == 0 {
		mode = "0"
	}

	err := l.sendVerify("SetPower", "{\"id\":0,\"method\":\"set_power\",\"params\":[\"%v\", \"%v\", %v, %v]}", power, effect, duration, mode)
	if err != nil {
		return err
	}

	l.stateMutex.Lock()
	l.latestState.On = power == "on"
	l.latestState.Color_Mode, _ = ColorModeFromString(mode)
	l.stateMutex.Unlock()
	return nil
}

func (l *Light) Toggle() error {
	err := l.sendVerify("Toggle", "{\"id\":0,\"method\":\"toggle\",\"params\":[]}")
	if err != nil {
		return err
	}

	l.stateMutex.Lock()
	l.latestState.On = !l.latestState.On
	l.stateMutex.Unlock()
	return nil
}

func (l *Light) SetDefault() error {
	panic("not implemented")
}

/*
"count" is the total number of visible state changing before color flow stopped.

	0 means infinite loop on the state changing.

"action" is the action taken after the flow is stopped.

	0 means smart LED recover to the state before the color flow started.
	1 means smart LED stay at the state when the flow is stopped.
	2 means turn off the smart LED after the flow is stopped.

"flow_expression" is the expression of the state changing series

From Yeelight's Inter-operation Specification
*/
func (l *Light) StartCf(count uint64, action uint8, flow_expression string) error {
	if action > 2 {
		return fmt.Errorf("StartCf() failed: action out of range")
	}

	err := l.sendVerify("StartCf", "{\"id\":0,\"method\":\"start_cf\",\"params\":[%v, %v, \"%v\"]}", count, action, flow_expression)
	if err != nil {
		return err
	}

	l.stateMutex.Lock()
	l.latestState.Flowing = true
	l.latestState.Flow_Params = flow_expression
	return nil
}

func (l *Light) StopCf() error {
	err := l.sendVerify("StopCf", "{\"id\":0,\"method\":\"stop_cf\",\"params\":[]}")
	if err != nil {
		return nil
	}

	l.stateMutex.Lock()
	l.latestState.Flowing = false
	l.latestState.Flow_Params = ""
	l.stateMutex.Unlock()
	return nil
}

/*
"class" can be "color", "hsv", "ct", "cf", "auto_dealy_off".

		"color" means change the smart LED to specified color and brightness.
	    "hsv" means change the smart LED to specified color and brightness.
	    "ct" means change the smart LED to specified ct and brightness.
	    "cf" means start a color flow in specified fashion.
	    "auto_delay_off" means turn on the smart LED to specified brightness and
		start a sleep timer to turn off the light after the specified minutes.

"val1", "val2", "val3" are class specific.

From Yeelight's Inter-operation Specification
*/
func (l *Light) SetScene(scene string, val1 string, val2 string, val3 string) error {
	panic("not implemented")
}

/*
"type2" currently can only be 0. (means power off)
"value" is the length of the timer (in minutes).

# From Yeelight's Inter-operation Specification

"type" was omited since type is a reserved keyword in Go, and it doesn't even serve any purpose, because it can only be set to 0
*/
func (l *Light) CronAdd(value string) error {
	panic("not implemented")
}

/*
"type2" the type of the cron job. (currently only support 0)

# From Yeelight's Inter-operation Specification

"type" was omited since type is a reserved keyword in Go, and it doesn't even serve any purpose, because it can only be set to 0
*/
func (l *Light) CronGet(type2 string) error {
	panic("not implemented")
}

/*
"type2" the type of the cron job. (currently only support 0)

# From Yeelight's Inter-operation Specification

"type" was omited since type is a reserved keyword in Go, and it doesn't even serve any purpose, because it can only be set to 0
*/
func (l *Light) CronDel(type2 string) error {
	panic("not implemented")
}

/*
		"action" the direction of the adjustment. The valid value can be:
			“increase": increase the specified property
			“decrease": decrease the specified property
			“circle": increase the specified property, after it reaches the max
			value, go back to minimum value

		"prop" the property to adjust. The valid value can be:
	        “bright": adjust brightness.
			“ct": adjust color temperature.
	        “color": adjust color. (When “prop" is “color", the “action" can only
			be “circle", otherwise, it will be deemed as invalid request.)

		From Yeelight's Inter-operation Specification
*/
func (l *Light) SetAdjust(action string, prop string) {
	panic("not implemented")
}

/*
		"action" the action of SetMusic command. The valid value can be:
	        0: turn off music mode.
	        1: turn on music mode.
	    "host" the IP address of the music server.
	    "port" the TCP port music application is listening on.
*/
func (l *Light) SetMusic(action string, host string, port string) {
	panic("not implemented")
}

/*
"name" the name of the device.
*/
func (l *Light) SetName(name string) {
	panic("not implemented")
}

/*
"ct_value" is the target color temperature. The type is integer and
range is 1700 ~ 6500 (k).

"effect" support two values: "sudden" and "smooth".

	If effect is "sudden",  then the color temperature will be changed directly to target value,
	under this case, the third parameter "duration" is ignored.
	If effect is "smooth", then the color temperature will be changed to target value in a
	gradual fashion, under this case, the total time of gradual change is specified in third
	parameter "duration".

"duration" specifies the total time of the gradual changing.

	The minimum support duration is 30 milliseconds.

From Yeelight's Inter-operation Specification
*/
func (l *Light) BgSetCtAbx(ct_value uint, effect string, duration string) error {
	if ct_value < 1700 || ct_value > 6500 {
		return fmt.Errorf("BgSetCtAbx() failed: ct_value out of range")
	}
	if effect != "sudden" && effect != "smooth" {
		return fmt.Errorf("BgSetCtAbx() failed: effect must be 'sudden' or 'smooth'")
	}

	durationConv, err := strconv.Atoi(duration)
	if err != nil {
		return fmt.Errorf("BgSetCtAbx() failed: duration must be an integer")
	}
	if durationConv < 30 {
		return fmt.Errorf("BgSetCtAbx() failed: duration must be at least 30 ms")
	}

	err = l.sendVerify("BgSetCtAbx", "{\"id\":0,\"method\":\"bg_set_ct_abx\",\"params\":[%v, \"%v\", %v]}", ct_value, effect, duration)
	if err != nil {
		return err
	}

	l.stateMutex.Lock()
	l.latestState.Bg_Ct = uint16(ct_value)
	l.stateMutex.Unlock()
	return nil
}

/*
"rgb_value" is the target color, whose type is integer.

	It should be expressed in decimal integer ranges from 0 to 16777215 (hex: 0xFFFFFF).
	RED, GREEN, BLUE = 0 up to 255
	rgb_value = RED*65536 + GREEN*256 + BLUE

"effect" support two values: "sudden" and "smooth".

	If effect is "sudden",  then the color temperature will be changed directly to target value,
	under this case, the third parameter "duration" is ignored.
	If effect is "smooth", then the color temperature will be changed to target value in a
	gradual fashion, under this case, the total time of gradual change is specified in third
	parameter "duration".

"duration" specifies the total time of the gradual changing.

	The minimum support duration is 30 milliseconds.

From Yeelight's Inter-operation Specification
*/
func (l *Light) BgSetRGB(rgb_value uint32, effect string, duration string) error {
	if rgb_value > 16777215 {
		return fmt.Errorf("SetRGB() failed: rgb_value out of range")
	}

	err := l.sendVerify("BgSetRGB", "{\"id\":0,\"method\":\"bg_set_rgb\",\"params\":[%v, \"%v\", %v]}", rgb_value, effect, duration)
	if err != nil {
		return err
	}

	l.stateMutex.Lock()
	l.latestState.Bg_RGB = rgb_value
	l.stateMutex.Unlock()
	return nil
}

/*
"hue" is the target hue value, whose type is integer.

	It should be expressed in decimal integer ranges from 0 to 359.

"sat" is the target saturation value whose type is integer.

	It's range is 0 to 100.

"effect" support two values: "sudden" and "smooth".

	If effect is "sudden",  then the color temperature will be changed directly to target value,
	under this case, the third parameter "duration" is ignored.
	If effect is "smooth", then the color temperature will be changed to target value in a
	gradual fashion, under this case, the total time of gradual change is specified in third
	parameter "duration".

"duration" specifies the total time of the gradual changing.

	The minimum support duration is 30 milliseconds.

From Yeelight's Inter-operation Specification
*/
func (l *Light) BgSetHSV(hue uint16, sat uint8, effect string, duration string) error {
	if hue > 359 {
		return fmt.Errorf("SetHSV() failed: hue out of range")
	}
	if sat > 100 {
		return fmt.Errorf("SetHSV() failed: sat out of range")
	}

	err := l.sendVerify("BgSetHSV", "{\"id\":0,\"method\":\"bg_set_hsv\",\"params\":[%v, %v, \"%v\", %v]}", hue, sat, effect, duration)
	if err != nil {
		return err
	}

	l.stateMutex.Lock()
	l.latestState.Bg_Hue = hue
	l.stateMutex.Unlock()
	return nil
}

/*
		"power" can only be "on" or "off".
			"on" means turn on the smart LED,
	        "off" means turn off the smart LED.

		"effect" support two values: "sudden" and "smooth".
			If effect is "sudden", then the color temperature will be changed directly to target value,
			under this case, the third parameter "duration" is ignored.
			If effect is "smooth", then the color temperature will be changed to target value in a
			gradual fashion, under this case, the total time of gradual change is specified in third
			parameter "duration".

		"duration" specifies the total time of the gradual changing. The minimum support duration is
		30 milliseconds.

		"mode" (optional):
			0: Normal turn on operation (default value)
			1: Turn on and switch to CT mode.
			2: Turn on and switch to RGB mode.
			3: Turn on and switch to HSV mode.
			4: Turn on and switch to color flow mode.
	        5: Turn on and switch to Night light mode. (Ceiling light only).

		From Yeelight's Inter-operation Specification
*/
func (l *Light) BgSetPower(power string, effect string, duration string, mode string) error {
	if len(mode) == 0 {
		mode = "0"
	}

	err := l.sendVerify("BgSetPower", "{\"id\":0,\"method\":\"bg_set_power\",\"params\":[\"%v\", \"%v\", %v, %v]}", power, effect, duration, mode)
	if err != nil {
		return err
	}

	l.stateMutex.Lock()
	l.latestState.Bg_On = power == "on"
	l.latestState.Bg_Color_Mode, _ = ColorModeFromString(mode)
	l.stateMutex.Unlock()
	return nil
}

/*
"brightness" is the target brightness.

	The type is integer and ranges from 1 to 100. The brightness is a
	percentage instead of a absolute value. 100 means maximum brightness
	while 1 means the minimum brightness.

"effect" support two values: "sudden" and "smooth".

	If effect is "sudden",  then the color temperature will be changed directly to target value,
	under this case, the third parameter "duration" is ignored.
	If effect is "smooth", then the color temperature will be changed to target value in a
	gradual fashion, under this case, the total time of gradual change is specified in third
	parameter "duration".

"duration" specifies the total time of the gradual changing.

	The minimum support duration is 30 milliseconds.

From Yeelight's Inter-operation Specification
*/
func (l *Light) BgSetBright(brightness uint8, effect string, duration string) error {
	if brightness < 1 || brightness > 100 {
		return fmt.Errorf("SetBright() failed: brightness out of range")
	}

	err := l.sendVerify("BgSetBright", "{\"id\":0,\"method\":\"bg_set_bright\",\"params\":[%v, \"%v\", %v]}", brightness, effect, duration)
	if err != nil {
		return err
	}

	l.stateMutex.Lock()
	l.latestState.Bg_Bright = brightness
	l.stateMutex.Unlock()
	return nil
}
